skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+"""_max_tree.pyx - building a max-tree from an image.
+
+This is an implementation of the max-tree, which is a morphological
+representation of the image. Many morphological operators can be built
+from this representation, namely attribute openings and closings.
+
+This file also contains implementations of max-tree based filters and
+functions to characterize the tree components. 
+"""
+
+import numpy as np
+
+from libc.stdlib cimport free, malloc, realloc
+from scipy.constants.constants import point
+from cProfile import label
+
+cimport numpy as np
+cimport cython
+
+ctypedef np.float64_t DTYPE_FLOAT64_t
+ctypedef np.int32_t DTYPE_INT32_t
+ctypedef np.uint32_t DTYPE_UINT32_t
+ctypedef np.uint64_t DTYPE_UINT64_t
+ctypedef np.int64_t DTYPE_INT64_t
+ctypedef np.uint8_t DTYPE_BOOL_t
+ctypedef np.uint8_t DTYPE_UINT8_t
+ctypedef np.int8_t DTYPE_INT8_t
+
+ctypedef fused dtype_t:
+    np.uint8_t
+    np.uint16_t
+    np.uint32_t
+    np.uint64_t
+    np.int8_t
+    np.int16_t
+    np.int32_t
+    np.int64_t
+    np.float32_t
+    np.float64_t
+
+cdef inline DTYPE_UINT64_t uint64_max(DTYPE_UINT64_t a, DTYPE_UINT64_t b): return a if a >= b else b
+cdef inline DTYPE_UINT64_t uint64_min(DTYPE_UINT64_t a, DTYPE_UINT64_t b): return a if a <= b else b
+
+
+# recursive function to get the root of the current tree. Importantly, the
+# function changes the tree (path compression), that reduces the complexity
+# from O(n*n) to O(n*log(n)).
+cdef DTYPE_INT64_t find_root(DTYPE_INT64_t[::1] parent, DTYPE_INT64_t index):
+    if parent[index] != index:
+        parent[index] = find_root(parent, parent[index])
+    return parent[index]
+
+# canonicalize generates a max-tree for which every node's
+# parent is a canonical node. Namely, either the representative node of the
+# same level or the representative node at the level below.
+cdef void canonize(dtype_t[::1] image, DTYPE_INT64_t[::1] parent,
+                   DTYPE_INT64_t[::1] sorted_indices):
+    cdef DTYPE_INT64_t q = 0
+    for p in sorted_indices:
+        q = parent[p]
+        if image[q] == image[parent[q]]:
+            parent[p] = parent[q]
+    return
+
+
+# helper function: a list of offsets is transformed to a list of points,
+# which are stored in a numpy array: rows correspond to different points,
+# columns to the dimensions (typically 2 or 3, but there is no limitation).
```

Comment:
Can you explain this differently? I don't really understand what this is doing. Why isn't it just a call to np.unravel_index?

... ah, I guess the *only* reason is that np.unravel_index doesn't support negative indices?

Maybe I would rename this function `unravel_offsets`? And note in the docstring the analogy to `unravel_index`.

Also, these should all be docstrings rather than comments.

Reply:
Exactly: offsets can be negative. 
Probably, this can be solved more elegantly?

Reply:
Yes, by contributing negative offsets to NumPy. 😂 This function is a good enough workaround for the moment. 😊